### PR TITLE
Unit tests for SecClient, improve error handling

### PR DESCRIPTION
**Pull request by smklein**
- Adds simple unit tests for a node with a single saga
- Improves error handling around edge-case behavior which previously panicked
- Cleaned up "trips" example slightly
- Clarify language in documentation around saga "creation" vs "starting"

Fixes https://github.com/oxidecomputer/steno/issues/13

### CODE REVIEW

File: src/sec.rs [hunk 1]
```diff
@@ -974,6 +973,16 @@ impl Sec {
                 waiter: done_tx,
             },
         };
+        if self.sagas.get(&saga_id).is_some() {
```
Comment (davepacheco) [inline src/sec.rs:976]: It looks like this changes this condition from an operational error to a programmer error.  Is that because the caller doesn't necessarily know if this saga was inserted already?

File: examples/trip.rs [hunk 2]
```diff
@@ -1,12 +1,10 @@
-//!
```
Comment (davepacheco) [inline examples/trip.rs:1]: Ha -- I had learned at one point the style rule that, for multi-line line comments, always include a blank line before and after.  I'm not sure where this came from.  It seems like we don't prefer that so I'll stop doing it.

**Review by davepacheco — COMMENTED** [review]
Thanks for the cleanup and tests!  (Obviously, steno is lacking a *lot* in tests.)

File: src/sec.rs (hunk 1 above)
Comment (smklein) [reply src/sec.rs:976]: So steno is very powerful, and presents an API for doing something complex - but from a usability perspective, I've bumped into a couple minor issues:

1. The interface has a lot of expectations that are validated at runtime, not compile time (this is one of them). This isn't "bad" at all - it's the right choice for many of these checks - but it delays when (as a user of the library) I get feedback on "has it been used correctly".
2. For many of those correctness checks, assertions are thrown to indicate programmer error. This is somewhat of a bummer, since we have not enumerated what invocations can cause panics to occur in the user-facing documentation. I stumbled onto this error by experimentation; I genuinely did not know what would happen when trying to "resume" a "created" saga.
3. Additionally, because of the "panic=abort" inclination and use of multiple tokio tasks, a single panic within the guts of steno (often many passed-messages away from the client usage) causes the entire process to abort, and unwinding "which client call caused this failure" is non-trivial. In this case, I saw "an assertion triggered in saga_insert", but had to trace back manually to uncover that was caused by a call to "saga_resume" in a test. In "real" code, this process is even more complex.

IMO, by moving many of the "things that are assertions but can be triggered by safe client API calls" from panics into errors, we'll get the following benefits:

- Way clearer where the "bad" invocation comes from; feedback can be delivered to callers without crashing the process.
- We can add tests for these edge cases! Using a "should_panic"-style test for an assertion that occurs in another tokio task is *really* hard (I spent a couple hours trying last night... the complexity is not worth the effort).

This is a more general answer to the "should we panic or return an error" question, but IMO these are worthwhile reasons to prefer the non-crashing option. As a perk of reserving assertions for invariants that *should not* be reachable by safe user code (with the exception of an assertion that explicitly has an error message saying something like "you called ____ wrong in the API, do ___ instead"), any assertion from steno has a wayyyy better signal/noise ratio on being a real bug.

**Review by davepacheco — APPROVED** [review]

File: src/sec.rs (hunk 1 above)
Comment (andrewjstone) [reply src/sec.rs:976]: Helpful discussion for future reference. Thanks!
